resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+
+If `N` is an expression, and the following rules specify the values to be
+assigned to `true(N)` and `false(N)`, but do not specify values for `null(N)`,
+`notNull(N)`, or `after(N)`, then they are by default assigned as follows:
+  - `null(N) = exit(after(N))`.
+  - `notNull(N) = join(true(N), false(N))`.
+  - `after(N) = notNull(N)`.
+
+If `N` is an expression, and the above rules specify the value to be assigned to
+`after(N)`, but do not specify values for `true(N)`, `false(N)`, `null(N)`, or
+`notNull(N)`, then they are all assigned the same value as `after(N)`.
+
+
+- **True literal**: If `N` is the literal `true`, then:
+  - Let `true(N) = before(N)`.
+  - Let `false(N) = exit(before(N))`.
+
+- **False literal**: If `N` is the literal `false`, then:
+  - Let `true(N) = exit(before(N))`.
+  - Let `false(N) = before(N)`.
+
+- **Shortcut and**: If `N` is a shortcut "and" expression of the form `E1 && E2`,
+  then:
+  - Let `before(E1) = before(N)`.
+  - Let `before(E2) = split(true(E1))`.
+  - Let `true(N) = drop(true(E2))`.
+  - Let `false(N) = join(split(false(E1)), false(E2))`.
+
+- **Shortcut or**: If `N` is a shortcut "or" expression of the form `E1 || E2`,
+  then:
+  - Let `before(E1) = before(N)`.
+  - Let `before(E2) = split(false(E1))`.
+  - Let `false(N) = drop(false(E2))`.
+  - Let `true(N) = join(split(true(E1)), true(E2))`.
+
+- **If-null**: If `N` is an if-null expression of the form `E1 ?? E2`, then:
+  - Let `before(E1) = before(N)`.
+  - Let `before(E2) = split(null(E1))`.
+  - Let `null(N) = drop(null(E2))`.
+  - Let `notNull(N) = join(split(notNull(E1)), notNull(E2))`.
+
+- **operator==** TODO
+
+- **Binary operator**: All binary operators other than `&&`, `||`, and `??` are
+  handled as calls to the appropriate `operator` method.
+
+- **Conditional expression**: If `N` is a conditional expression of the form `E1
+  ? E2 : E3`, then:
+  - Let `before(E1) = before(N)`.
+  - Let `before(E2) = split(true(E1))`.
+  - Let `before(E3) = split(false(E1))`.
+  - Let `after(N) = join(after(E2), after(E3))`.
+  - Let `true(N) = join(true(E2), true(E3))`.
+  - Let `false(N) = join(false(E2), false(E3))`.
+  - Let `null(N) = join(null(E2), null(E3))`.
+  - Let `notNull(N) = join(notNull(E2), notNull(E3))`.
+
+
+- **Local variable assignment**: If `N` is an expression of the form `x = E1`
+  where `x` is a local variable, then:
+  - Let `before(E1) = before(N)`.
+  - Let `after(N) = assign(x, E1, after(E1))`.
+  - Let `true(N) = assign(x, E1, true(E1))`.
+  - Let `false(N) = assign(x, E1, false(E1))`.
+  - Let `null(N) = assign(x, E1, null(E1))`.
+  - Let `notNull(N) = assign(x, E1, notNull(E1))`.
+
+
+
+### Statements
+
+- **Conditional statement**: If `N` is a conditional statement of the form `if
+  (E) S1 else S2` then:
+  - Let `before(E) = before(N)`.
+  - Let `before(S1) = split(true(E))`.
+  - Let `before(S2) = split(false(E))`.
+  - Let `after(N) = join(after(S1), after(S2))`.
+  - Let `true(N) = join(true(S1), true(S2))`.
```

Comment:
Since statements don't have values, we shouldn't need `true(N)`, `false(N)`, `null(N)`, or `notNull(N)`.  Just `after(N)` is sufficient.